### PR TITLE
Optimize bit count polyfills

### PR DESCRIPTION
**Pull request by MaxGraey**

### CODE REVIEW

File: src/support/bits.cpp [hunk 1]
```diff
@@ -42,7 +42,7 @@ template<> int PopCount<uint32_t>(uint32_t v) {
 }
 
 template<> int PopCount<uint64_t>(uint64_t v) {
-  return PopCount((uint32_t)v) + PopCount((uint32_t)(v >> 32));
+  return PopCount((uint32_t)v) + (v >> 32 ? PopCount((uint32_t)(v >> 32)) : 0);
```
Comment (tlively) [inline src/support/bits.cpp:45]: What is the motivation for this change? This seems slightly harder to read than before.

File: src/support/bits.h [hunk 2]
```diff
@@ -65,7 +65,9 @@ template<typename T> int CountTrailingZeroes(T v) {
 template<typename T> int CountLeadingZeroes(T v) {
   return CountLeadingZeroes(typename std::make_unsigned<T>::type(v));
 }
-template<typename T> bool IsPowerOf2(T v) { return v != 0 && PopCount(v) == 1; }
+template<typename T> bool IsPowerOf2(T v) {
+  return v != 0 && (v & (v - 1)) == 0;
+}
```
Comment (tlively) [inline src/support/bits.h:70]: Nice 👍 I like that this lets us delete code.

Comment (MaxGraey) [reply src/support/bits.h:70]: usually `popcnt(x) == 1` (btw it include x != 0 case also) faster but only if we have native popcnt support. Alos LLVM & GCC smart enought to replace pattern above to popcnt(x) == 1 if it possible

File: src/support/bits.cpp (hunk 1 above)
Comment (MaxGraey) [reply src/support/bits.cpp:45]: Main motivation is consistency with CountLeadingZeros for 64-bit and potentially speedup calculation when high part of 64-bit is zero

File: src/support/bits.cpp [hunk 3]
```diff
@@ -54,8 +56,8 @@ template<> int PopCount<uint32_t>(uint32_t v) {
 }
 
 template<> int PopCount<uint64_t>(uint64_t v) {
-#if __has_builtin(__builtin_popcountll) || defined(__GNUC__)
-  return __builtin_popcountll(v);
+#if __has_builtin(__builtin_popcount) || defined(__GNUC__) || defined(_MSC_VER)
```
Comment (MaxGraey) [inline src/support/bits.cpp:59]: use `__has_builtin(__builtin_popcount)` instead `__has_builtin(__builtin_popcountll)` due to clang-format forcing line terminator and carry `defined(_MSC_VER)` to new line and this looks weird.

File: src/support/bits.h [hunk 4]
```diff
@@ -65,17 +65,21 @@ template<typename T> int CountTrailingZeroes(T v) {
...
 
 template<typename T, typename U> inline static T RotateLeft(T val, U count) {
-  T mask = sizeof(T) * CHAR_BIT - 1;
+  auto value = typename std::make_unsigned<T>::type(val);
```
Comment (MaxGraey) [inline src/support/bits.h:73]: pretty important cast to `unsigned`. Otherwise not LLVM not GCC can't fold this to `rol` / `rot`

File: src/support/bits.cpp (hunk 1 above)
Comment (tlively) [reply src/support/bits.cpp:62]: I think the consistency with CountLeadingZeroes is a non-goal here because they are fundamentally different optimizations. If we could rewrite CountLeadingZeroes to match PopCount, I think we would want to do that for improved readability.

If we're going to sacrifice some readability for performance, it would be good to see that the performance difference is measurable rather than hypothetical.

This PR LGTM other than this point, so it might be nice to split this out and land the rest.

Comment (MaxGraey) [reply src/support/bits.cpp]: Alright. I switched to prev implementation for PopCount<uint64_t>
